saleor/graphql/product/mutations/channels.py

```diff
@@ -156,12 +171,103 @@ def add_channels(cls, product: "ProductModel", add_channels: List[Dict]):
                     defaults["available_for_purchase"] = datetime.date.today()
                 else:
                     defaults["available_for_purchase"] = available_for_purchase_date
-            ProductChannelListing.objects.update_or_create(
+            product_channel_listing, _ = ProductChannelListing.objects.update_or_create(
                 product=product, channel=channel, defaults=defaults
             )
+            cls.validate_variants(add_variants, remove_variants)
```

Comment:
IMO it should be called in some `clean` method, not here, as `add_channels` is called in `save`, which assumed that data are already cleaned.